include/DeterministicNoiseSimulator.hpp

```diff
@@ -3,124 +3,132 @@
-    }
-
-    DeterministicNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const std::string& noise_effects, double noise_prob, double amp_noise_prob, double multiQubitGateFactor):
-        qc(qc) {
-        setNoiseEffects(noise_effects);
-        initializeNoiseProbabilities(noise_prob, amp_noise_prob, multiQubitGateFactor);
-    }
-
-    double noiseProb                 = 0.0;
-    double ampDampingProb            = 0.0;
-    double noiseProbSingleQubit      = 0.0;
-    double ampDampingProbSingleQubit = 0.0;
-    double noiseProbMultiQubit       = 0.0;
-    double ampDampingProbMultiQubit  = 0.0;
+    explicit DeterministicNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, unsigned long long seed = 0):
+        DeterministicNoiseSimulator(qc, std::string("APD"), 0.001, std::optional<double>{}, 2, false, seed) {}
+
+    DeterministicNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc,
+                                const std::string&                       cNoiseEffects,
+                                double                                   cNoiseProbability,
+                                std::optional<double>                    cAmpDampingProbability,
+                                double                                   cMultiQubitGateFactor,
+                                bool                                     unoptimizedSim = false,
+                                unsigned long long                       seed           = 0):
+        Simulator<DDPackage>(seed),
+        qc(qc),
+        noiseProbSingleQubit(cNoiseProbability),
+        noiseProbMultiQubit(cNoiseProbability * cMultiQubitGateFactor),
+        sequentiallyApplyNoise(unoptimizedSim),
+        useDensityMatrixType(!unoptimizedSim) {
+        // setNoiseEffects
+        for (auto noise: cNoiseEffects) {
+            switch (noise) {
+                case 'A':
+                    noiseEffects.push_back(dd::amplitudeDamping);
+                    break;
+                case 'P':
+                    noiseEffects.push_back(dd::phaseFlip);
+                    break;
+                case 'D':
+                    noiseEffects.push_back(dd::depolarization);
+                    break;
+                case 'I':
+                    noiseEffects.push_back(dd::identity);
+                    break;
+                default:
+                    throw std::runtime_error("Unknown noise operation '" + cNoiseEffects + "'\n");
+            }
+        }
 
-    void initializeNoiseProbabilities(double cGateNoiseProbability, double cAmplitudeDampingProb = -1, double cMultiQubitGateFactor = 2) {
-        noiseProb            = cGateNoiseProbability;
-        noiseProbSingleQubit = cGateNoiseProbability;
-        noiseProbMultiQubit  = cGateNoiseProbability * cMultiQubitGateFactor;
+        // initializeNoiseProbabilities
+        noiseProbSingleQubit = cNoiseProbability;
+        noiseProbMultiQubit  = cNoiseProbability * cMultiQubitGateFactor;
```

Comment:
these are already initialised at the top of this function

Reply:
Removed redundant lines